library/std/src/io/mod.rs

```diff
@@ -1385,17 +1385,18 @@ pub trait Write {
     ///
     /// This function will attempt to write the entire contents of `buf`, but
     /// the entire write might not succeed, or the write may also generate an
-    /// error. A call to `write` represents *at most one* attempt to write to
+    /// error. Typically, a call to `write` represents one attempt to write to
     /// any wrapped object.
     ///
     /// Calls to `write` are not guaranteed to block waiting for data to be
     /// written, and a write which would otherwise block can be indicated through
     /// an [`Err`] variant.
     ///
-    /// If the return value is [`Ok(n)`] then it must be guaranteed that
-    /// `n <= buf.len()`. A return value of `0` typically means that the
-    /// underlying object is no longer able to accept bytes and will likely not
-    /// be able to in the future as well, or that the buffer provided is empty.
+    /// If the return value is [`Ok(n)`] then it must be guaranteed that `n <=
+    /// buf.len()`. Unless `input` is empty, this function shouldn’t return `0`
+    /// since caller may interpret that as an error (the default implementation
+    /// of [`Write::write_all`] does exactly that).  To indicate lack of space
+    /// function should return [`ErrorKind::StorageFull`] error instead.
```

Comment:
Id's say something like: "Unless `input` is empty, this function should only return `Ok(0)` if the device is in an EoF state, indicating that no more data will be accepted and the caller shouldn't retry the write."

Reply:
That seems weird to me.  Why would we use `Ok` to indicate an error?

Reply:
That conversation was more about what the client should do with Ok(0) return value.  (Or at least that’s how I interpret it and the final decision ‘that the current behaviour of write_all for Ok(0) is the right one’).  The question here remains whether write should ever return Ok(0) (provided non-empty input).

My position is that it shouldn’t and it was defect of the initial design to allow it.  POSIX agrees.  (I think Windows does too but honestly I couldn’t find clear confirmation in documentation).  We’re [now (about to) have ErrorKind::StorageFull](https://github.com/rust-lang/rust/issues/86442) thus it is possible to rectify the situation and direct implementers to the correct ErrorKind they should be using in case of full storage.  It is also worth noting that (at least on Linux) std::fs::File returns an error rather than returning Ok(0).

Reply:
I think that's fair; an implementation could choose to return `StorageFull` or `WriteZero` instead of `Ok(0)`. I think it still make sense that *if* a client returns `Ok(0)`, *then* it indicates an EoF state (to prevent infinite loops).